### PR TITLE
I have added a class name GBBuilder to the factory.py file to generat…

### PR DESCRIPTION
**Pull request by usaikia**
…e grain boundary structures using amisgb code

### CODE REVIEW

File: pyiron_atomistics/atomistics/structure/factory.py [hunk 1]
```diff
@@ -332,3 +332,65 @@ def element(parent_element, new_element_name=None, spin=None, potential_file=Non
                 parent_element=parent_element, new_element=new_element_name
             )
         return periodic_table.element(new_element_name)
+
+# importing the modules required by aimsgb
+from pyiron.atomistics.structure.atoms import pyiron_to_pymatgen, pymatgen_to_pyiron
+from pymatgen import Structure, Lattice, PeriodicSite
+from aimsgb import GrainBoundary, Grain, GBInformation
```
Comment (jan-janssen) [inline pyiron_atomistics/atomistics/structure/factory.py]: Please move the importing of classes to the top of the file.

File: pyiron_atomistics/atomistics/structure/factory.py [hunk 2]
```diff
@@ -332,3 +332,65 @@ def element(parent_element, new_element_name=None, spin=None, potential_file=Non
...
+
+class GBBuilder:
```
Comment (jan-janssen) [inline pyiron_atomistics/atomistics/structure/factory.py]: Do we need a separate class here?

File: pyiron_atomistics/atomistics/structure/factory.py [hunk 3]
```diff
@@ -332,3 +332,65 @@ def element(parent_element, new_element_name=None, spin=None, potential_file=Non
...
+
+    def gb_info(axis, max_sigma):
```
Comment (jan-janssen) [inline pyiron_atomistics/atomistics/structure/factory.py]:
```suggestion
    @staticmethod
    def gb_info(axis, max_sigma):
```

File: pyiron_atomistics/atomistics/structure/factory.py [hunk 4]
```diff
@@ -332,3 +332,65 @@ def element(parent_element, new_element_name=None, spin=None, potential_file=Non
...
+        print("{sigma value: {'theta': [theta value],")
+        print("'plane': the GB plane")
+        print("'rot_matrix': the rotational matrix,")
+        print("'csl': the csl matrix}}\n")
+        print("To construct the grain boundary select a GB plane and sigma value from the list below and \
+        \npass it to the GBBuilder.gb_build() function along with the rotational axis and initial bulk structure.")
+
+        return GBInformation(axis=axis, max_sigma=max_sigma)
+
+    def gb_build(axis, sigma, plane, initial_struct, to_primitive=False,
```
Comment (jan-janssen) [inline pyiron_atomistics/atomistics/structure/factory.py]:
```suggestion
    @staticmethod
    def gb_build(axis, sigma, plane, initial_struct, to_primitive=False,
```

File: pyiron_atomistics/atomistics/structure/factory.py [hunk 5]
```diff
@@ -332,3 +332,65 @@ def element(parent_element, new_element_name=None, spin=None, potential_file=Non
...
+
+        Args:
+            axis : Rotational axis for the GB you want to construct (for example, axis=[1,0,0])
+            max_sigma (int) : The maximum value of sigma upto which you want to consider for your
+            GB (for example, sigma=5)
+
+        Returns:
+            A list of possible GB structures.
+        """
+        print("Possible GB structures for rotational axis",axis,"and sigma cutoff", max_sigma,"are listed in the format:\n")
```
Comment (pmrv) [inline pyiron_atomistics/atomistics/structure/factory.py]: This can go into the docstring.

**Comment by jan-janssen** [comment]
@usaikia Thank you very much, I guess there are some minor fixes required but it is really great to see more and more people contributing to pyiron.

File: pyiron_atomistics/atomistics/structure/factory.py (hunk 4 above)
Comment (usaikia) [reply pyiron_atomistics/atomistics/structure/factory.py]: Thank you for the suggestions. I will modify the code accordingly. I am not sure about the class. Should I remove it or keep it. I thought may be we can add more things to the GBBuilder class in the future

File: pyiron_atomistics/atomistics/structure/factory.py (hunk 5 above)
Comment (usaikia) [reply pyiron_atomistics/atomistics/structure/factory.py]: @pmrv thank you. I will add them to the docstring.

File: pyiron_atomistics/atomistics/structure/factory.py [hunk 6]
```diff
@@ -32,6 +32,10 @@
 from pyiron_atomistics.atomistics.structure.periodic_table import PeriodicTable
 from pyiron_base import Settings, PyironFactory
 import types
+# modules required for aimsgb
```
Comment (liamhuber) [inline pyiron_atomistics/atomistics/structure/factory.py]:
```suggestion
```

**Review by liamhuber — CHANGES_REQUESTED** [review]
Very nice!

As Jan suggested, this probably does not to be a separate class. More critically, right now it's not accessible 'pyironically' -- i.e. without an explicit import. Somehow these functions need to be directly attached to the main factory class.

For now I would suggest renaming these to `aimsgb_info` and `aimsgb_build` and making them (static) methods of the `StructureFactory` directly. 
  We can always refactor these functions off into their own class later, but this sort of naming convention and sticking them on the main class is most consistent with (a) what we already do for other structure constructors and (b) my best sense of where consensus is landing on the discussion in [issue 1307](https://github.com/pyiron/pyiron/issues/1307).

**Comment by pmrv** [comment]
We should also add `aimsgb` to the list of depencies [here](https://github.com/pyiron/pyiron_atomistics/blob/master/setup.py).  @usaikia Is the module available on pip/conda?  How are you installing it currently?

**Comment by usaikia** [comment]
@pmrv  aimsgb is available in github. I downloaded it into my local directory and used pip install . to install it. Here is the link to the github page  https://github.com/ksyang2013/aimsgb

**Comment by pmrv** [comment]
> @pmrv aimsgb is available in github. I downloaded it into my local directory and used pip install . to install it. Here is the link to the github page https://github.com/ksyang2013/aimsgb

Ah, @jan-janssen already uploaded it to [PyPI](https://pypi.org/project/aimsgb/). There's no problem then adding it to setup.py.

**Comment by jan-janssen** [comment]
Please also add it in https://github.com/pyiron/pyiron_atomistics/blob/aimsgb_in_pyiron/.ci_support/environment.yml

File: pyiron_atomistics/atomistics/structure/factory.py [hunk 7]
```diff
@@ -332,3 +335,59 @@ def element(parent_element, new_element_name=None, spin=None, potential_file=Non
...
+                           to be deleted. The first t and second b from the left hand side represents
+                           the layers at the GB interface. Default value is delete_layer='0b0t0b0t', which
+                           means no deletion of layers.
+            add_if_dist : If you want to add extra interface distance, you can specify add_if_dist.
+                           Default value is add_if_dist=0.0
+            to_primitive : To generate primitive or non-primitive GB structure. Default value is
+                            to_primitive=False
+
+        Returns:
+            Grain boundary structure in pyiron structure file format.
```
Comment (pmrv) [inline pyiron_atomistics/atomistics/structure/factory.py]: For consistency with the methods explicitly state the class that will be returned.
```suggestion
            :class:`.Atoms`: final grain boundary structure
```

**Review by pmrv — APPROVED** [review]
Except for the small nit above, I'm happy.

**Review by jan-janssen — CHANGES_REQUESTED** [review]
Please still update the `environment.yml` and `setup.py` file to include the latest version of `pymatgen` and `aimsgb`

**Comment by usaikia** [comment]
> Please still update the `environment.yml` and `setup.py` file to include the latest version of `pymatgen` and `aimsgb`

@jan-janssen sorry for the inconvenience. I have no previous knowledge in writing code for a package. This is my first experience and I am not sure what you suggested me to do in this comment. Can you please explain what you want me to update and how?

**Review by jan-janssen — APPROVED** [review]
Looks good to me - I added the dependencies.

File: pyiron_atomistics/atomistics/structure/factory.py [hunk 8]
```diff
@@ -32,6 +32,9 @@
 from pyiron_atomistics.atomistics.structure.periodic_table import PeriodicTable
 from pyiron_base import Settings, PyironFactory
 import types
+from pyiron.atomistics.structure.atoms import pyiron_to_pymatgen, pymatgen_to_pyiron
```
Comment (jan-janssen) [inline pyiron_atomistics/atomistics/structure/factory.py]:
```suggestion
from pyiron_atomistics.atomistics.structure.atoms import pyiron_to_pymatgen, pymatgen_to_pyiron
```

**Comment by jan-janssen** [comment]
@liamhuber From my perspective the pull request is ready to go, can you take another look?

**Review by liamhuber — APPROVED** [review]
Super!

**Comment by jan-janssen** [comment]
@usaikia as discussed in the meeting today, feel free to merge this pull request any time.

**Comment by usaikia** [comment]
> @usaikia as discussed in the meeting today, feel free to merge this pull request any time.

done

**Pull request merged**